src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -565,92 +566,60 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 | DefKind::ForeignTy,
                 did,
             ) => {
-                debug!("looking for associated item named {} for item {:?}", item_name, did);
-                // Checks if item_name belongs to `impl SomeItem`
-                let assoc_item = tcx
-                    .inherent_impls(did)
-                    .iter()
-                    .flat_map(|&imp| {
-                        tcx.associated_items(imp).find_by_name_and_namespace(
-                            tcx,
-                            Ident::with_dummy_span(item_name),
-                            ns,
-                            imp,
-                        )
-                    })
-                    .map(|item| (item.kind, item.def_id))
-                    // There should only ever be one associated item that matches from any inherent impl
-                    .next()
-                    // Check if item_name belongs to `impl SomeTrait for SomeItem`
-                    // FIXME(#74563): This gives precedence to `impl SomeItem`:
-                    // Although having both would be ambiguous, use impl version for compatibility's sake.
-                    // To handle that properly resolve() would have to support
-                    // something like [`ambi_fn`](<SomeStruct as SomeTrait>::ambi_fn)
-                    .or_else(|| {
-                        let kind =
-                            resolve_associated_trait_item(did, module_id, item_name, ns, self.cx);
-                        debug!("got associated item kind {:?}", kind);
-                        kind
-                    });
-
-                if let Some((kind, id)) = assoc_item {
-                    let out = match kind {
-                        ty::AssocKind::Fn => "method",
-                        ty::AssocKind::Const => "associatedconstant",
-                        ty::AssocKind::Type => "associatedtype",
-                    };
-                    Some(if extra_fragment.is_some() {
-                        Err(ErrorKind::AnchorFailure(AnchorFailure::RustdocAnchorConflict(ty_res)))
+                if expect_field {
+                    self.resolve_variant_or_field(item_name, did, ty_res, extra_fragment)
+                } else {
```

Comment:
This function is already super long with lots of indentation - could you change this to a separate match arm? `Res::Def(...) if expect_field => { ... }`

Alternatively you could turn this whole match into a separate function, which lets you do an early return.